File: week_2/code/day_9/a_validate_bst.js
// Definition for a binary tree node.
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
    function valid(node, leftBoundary, rightBoundary) {
        if (!node) {
            return true;
        }

        if (node.val <= leftBoundary || node.val >= rightBoundary) {
            return false;
        }

        return (
            valid(node.left, leftBoundary, node.val) &&
            valid(node.right, node.val, rightBoundary)
        );
    }

    return valid(root, -Infinity, Infinity);
};

// Example Usage
// Valid BST
//   2
//  / \
// 1   3
const validRoot = new TreeNode(2, new TreeNode(1), new TreeNode(3));
console.log(isValidBST(validRoot)); // true

// Invalid BST
//   5
//  / \
// 1   4
//    / \
//   3   6
const invalidRoot = new TreeNode(5, new TreeNode(1), new TreeNode(4, new TreeNode(3), new TreeNode(6)));
console.log(isValidBST(invalidRoot)); // false

File: week_2/code/day_9/a_validate_bst.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def valid(node, left_boundary, right_boundary):
            if not node:
                return True
            
            # The current node's value must be strictly between the boundaries
            if not (left_boundary < node.val < right_boundary):
                return False
            
            # Left child must be strictly less than node.val
            # Right child must be strictly greater than node.val
            return (valid(node.left, left_boundary, node.val) and
                    valid(node.right, node.val, right_boundary))

        return valid(root, float("-inf"), float("inf"))

# Example Usage
if __name__ == "__main__":
    sol = Solution()
    
    # Valid BST
    #   2
    #  / \
    # 1   3
    valid_root = TreeNode(2, TreeNode(1), TreeNode(3))
    print(sol.isValidBST(valid_root)) # True
    
    # Invalid BST
    #   5
    #  / \
    # 1   4
    #    / \
    #   3   6
    invalid_root = TreeNode(5, TreeNode(1), TreeNode(4, TreeNode(3), TreeNode(6)))
    print(sol.isValidBST(invalid_root)) # False
